utils/frame/try-runtime/cli/src/lib.rs

```diff
@@ -15,23 +15,60 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//! `Structopt`-ready struct for `try-runtime`.
+//! `Structopt`-ready structs for `try-runtime`.
 
-use parity_scale_codec::Decode;
-use std::{fmt::Debug, path::PathBuf, str::FromStr};
+use parity_scale_codec::{Decode, Encode};
+use std::{fmt::Debug, path::PathBuf, str::FromStr, sync::Arc};
 use sc_service::Configuration;
 use sc_cli::{CliConfiguration, ExecutionStrategy, WasmExecutionMethod};
 use sc_executor::NativeExecutor;
 use sc_service::NativeExecutionDispatch;
 use sp_state_machine::StateMachine;
 use sp_runtime::traits::{Block as BlockT, NumberFor};
-use sp_core::storage::{StorageData, StorageKey, well_known_keys};
+use sp_core::{
+	offchain::{
+		OffchainWorkerExt, OffchainDbExt, TransactionPoolExt,
+		testing::{TestOffchainExt, TestTransactionPoolExt}
+	},
+	storage::{StorageData, StorageKey, well_known_keys},
+};
+use sp_keystore::{KeystoreExt, testing::KeyStore};
+use remote_externalities::{Builder, Mode, SnapshotConfig, OfflineConfig, OnlineConfig, rpc_api};
 
-/// Various commands to try out the new runtime, over configurable states.
-///
-/// For now this only assumes running the `on_runtime_upgrade` hooks.
-#[derive(Debug, structopt::StructOpt)]
-pub struct TryRuntimeCmd {
+#[derive(Debug, Clone, structopt::StructOpt)]
+pub enum Command {
+	OnRuntimeUpgrade(OnRuntimeUpgradeCmd),
+	OffchainWorker(OffchainWorkerCmd),
+}
+
+#[derive(Debug, Clone, structopt::StructOpt)]
+pub struct OnRuntimeUpgradeCmd {
+	#[structopt(subcommand)]
+	pub state: State,
+}
+
+#[derive(Debug, Clone, structopt::StructOpt)]
+pub struct OffchainWorkerCmd {
+	/// Hash of the block whose header to use to execute the offchain worker.
+	#[structopt(short, long, multiple = false, parse(try_from_str = parse::hash))]
+	pub header_at: String,
+
+	#[structopt(subcommand)]
+	pub state: State,
```

Comment:
A block number is kept in the state, and one in the header. You can compare the two and panic/warn in case of mismtach.

Reply:
Cleaning up header stuff in a follow up